dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1183 @@
+	cs.blockState = new(mocks.BlockState)
+	cs.blockState.(*mocks.BlockState).On("BestBlockHeader").Return(header, nil)
+	fin, err = types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, number-2,
+		types.NewDigest())
+	require.NoError(t, err)
+	cs.blockState.(*mocks.BlockState).On("GetHighestFinalisedHeader").Return(fin, nil)
+	cs.blockState.(*mocks.BlockState).On("GetHashByNumber", mock.AnythingOfType("uint")).Return(common.
+		Hash{}, nil)
+	cs.blockState.(*mocks.BlockState).On("HasHeader", mock.AnythingOfType("common.Hash")).Return(true, nil)
+
+	err = cs.setPeerHead(testPeer, hash, number-1)
+	require.NoError(t, err)
+	expected = &peerState{
+		who:    testPeer,
+		hash:   hash,
+		number: number - 1,
+	}
+	require.Equal(t, expected, cs.peerState[testPeer])
+	select {
+	case <-cs.workQueue:
+		t.Fatal("should not put fork we already have into work queue")
+	default:
+	}
+}
+
+func TestChainSync_sync_bootstrap_withWorkerError(t *testing.T) {
+	t.Parallel()
+
+	cs := newTestChainSync(t)
+	ctrl := gomock.NewController(t)
+	mockBlockState := NewMockBlockState(ctrl)
+	mockHeader, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, 0,
+		types.NewDigest())
+	require.NoError(t, err)
+	mockBlockState.EXPECT().BestBlockHeader().Return(mockHeader, nil).Times(2)
+	cs.blockState = mockBlockState
+	cs.handler = newBootstrapSyncer(mockBlockState)
+
+	mockNetwork := NewMockNetwork(ctrl)
+	startingBlock := variadic.MustNewUint32OrHash(1)
+	max := uint32(128)
+	mockNetwork.EXPECT().DoBlockRequest(peer.ID("noot"), &network.BlockRequestMessage{
+		RequestedData: 19,
+		StartingBlock: *startingBlock,
+		EndBlockHash:  nil,
+		Direction:     0,
+		Max:           &max,
+	})
+	cs.network = mockNetwork
+
+	go cs.sync()
+	defer cs.cancel()
+
+	testPeer := peer.ID("noot")
+	cs.peerState[testPeer] = &peerState{
+		number: 1000,
+	}
+
+	cs.workQueue <- cs.peerState[testPeer]
+
+	select {
+	case res := <-cs.resultQueue:
+		expected := &workerError{
+			err: errNilResponse, // since MockNetwork returns a nil response
+			who: testPeer,
+		}
+		require.Equal(t, expected, res.err)
+	case <-time.After(time.Second * 5):
+		t.Fatal("did not get worker response")
+	}
+
+	require.Equal(t, bootstrap, cs.state)
+}
+
+func TestChainSync_sync_tip(t *testing.T) {
+	t.Parallel()
+
+	cs := newTestChainSync(t)
+	cs.blockState = new(mocks.BlockState)
+	header, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, 1000,
+		types.NewDigest())
+	require.NoError(t, err)
+	cs.blockState.(*mocks.BlockState).On("BestBlockHeader").Return(header, nil)
+	cs.blockState.(*mocks.BlockState).On("GetHighestFinalisedHeader").Return(header, nil)
+
+	go cs.sync()
+	defer cs.cancel()
+
+	testPeer := peer.ID("noot")
+	cs.peerState[testPeer] = &peerState{
+		number: 999,
+	}
+
+	cs.workQueue <- cs.peerState[testPeer]
+	time.Sleep(time.Second)
```

Comment:
Is there a way to have this event driven?
Maybe by closing a signal channel in the mock calls? 🤔